src/primitives.rs

```diff
@@ -0,0 +1,133 @@
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
+use crate::requests::{BodyApiVersionRequest, BodyMeatdataApiVersionRequest};
+use bytes::Buf;
+use std::fmt::{Error, Formatter};
+
+#[derive(FromPrimitive)]
+pub enum ApiKey {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersion {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeError {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct HeaderRequest {
+    pub api_key: i16,
+    pub api_version: i16,
+    pub correlation_id: i32,
+    pub client_id: String,
+}
+
+impl std::fmt::Display for HeaderRequest {
+    fn fmt(&self, _: &mut Formatter) -> Result<(), Error> {
+        println!(
+            "api_key: {}, api_version: {}, correlation_id: {}, client_id: {}",
+            self.api_key, self.api_version, self.correlation_id, self.client_id
+        );
+        Ok(())
+    }
+}
+
+impl FromByte for HeaderRequest {
+    fn decode(buf: &mut Buf) -> Result<Self, DecodeError> {
+        Ok(HeaderRequest {
+            api_key: decode_buffer(buf)?,
+            api_version: decode_buffer(buf)?,
+            correlation_id: decode_buffer(buf)?,
+            client_id: decode_buffer(buf)?,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct HeaderResponse {
+    pub correlation_id: i32,
+}
+
+pub trait BodyRequest {}
+pub struct DecodedRequest {
+    pub size: i32,
+    pub header: HeaderRequest,
+    pub body: Box<dyn BodyRequest>,
```

Comment:
I'm not convinced using a boxed trait object here is the right approach. The type of a trait object is _erased_ --- users of this struct can _only_ interact with the trait object by calling methods on the trait (in this case, `BodyRequest`, which has no methods). This means that if I get a Kafka request and decode it to a `DecodedRequest`, I can only access the size and common header; I can't actually use any of the data in the request body.

As an alternative, I'd suggest replacing this with an `enum` type that enumerates the possible kind of request body. Something like this:
```rust
pub struct DecodedRequest {
    pub size: i32,
    pub header: RequestHeader,
    pub body: RequestBody,
}

pub enum RequestBody {
    ApiVersions(ApiVersionsRequest),
    Metadata(MetadataRequest),
    // ... other Kafka request types ...
}

pub struct ApiVersionsRequest {
    // ...
}

pub struct MetadataRequest {
    // ...
}

// and so on for all the request body types...
```

This way, as a user of this API, I can write code that has different behaviours based on what kind of request I'm given. For example:

```rust
fn process_kafka_request(request: DecodedRequest) {
    match request.body {
        RequestBody::ApiVersions(api_versions) => {
            // do functionality specific to API versions requests here...
        },
        RequestBody::Metadata(metadata) => {
            // do functionality specific to metadata requests here...
        },
        // ... and so on for the other Kafka request types ...
    }
}
```

I can _also_ write functions that take a specific type of request body. For example,
```rust
fn handle_api_metadata_request(req: &MetadataRequest) { 
    // ...
}
```
This function would take the Metadata request body type only. It could be called by code that destructures a `DecodedRequest` to that type.

Reply:
`enum` sounds much better than `box`. I was thinking about using the `header.api_key` to tell the type of the body, and it is indeed awkward.

Reply:
@hawkw I pushed one more commit to replace `box` with `enum`. It looks good.

Can you check the latest commit (https://github.com/ruig2/KafkaRustCodec/pull/2/commits/3cedb72d82a455d11e0e2f040cef93788c420569)? If everything fine then I can implement more requests and responses and add tests to see if this is what we want.

One more question is about the codec interface. Currently, we are using `buffer. Should `buffer` be the input to the codec? I don't think codec should handle TCP stream or control concurrency, instead, it is just decoding (not even encoding in our proxy scenario) a short buffer given by Linkerd-proxy.

And, is the buffer always **short** and contains only one or very few Kafka requests/responses? If the buffer is dynamic (as a data stream and new messages may arrive any time) or very large, then maybe we need to do something else. I don't know how Linkerd handles the buffer in case of Kafka traffic. My experience with Tokio is that every Kafka request and response can be converted into a separated (maybe paged as 4k?) buffer.

Reply:
> One more question is about the codec interface. Currently, we are using `buffer. Should `buffer` be the input to the codec? I don't think codec should handle TCP stream or control concurrency, instead, it is just decoding (not even encoding in our proxy scenario) a short buffer given by Linkerd-proxy.

Yeah, I think this is the right design; I would not expect the codec to be aware of TCP connections at all; simply decode bytes. In the future, I think we may want to be able to support asynchronously decoding bytes from a stream as they become available, using the `AsyncRead` trait. However, this is something we can probably build on top of a codec built around `Bytes` buffers.

> And, is the buffer always **short** and contains only one or very few Kafka requests/responses? If the buffer is dynamic (as a data stream and new messages may arrive any time) or very large, then maybe we need to do something else.

I think it's reasonable to expect the proxy to provide a buffer per request. If Kafka requests can have streaming bodies of arbitrary length (I'm not sure if this exists in the wire protocol or not?) we probably only want to parse the request headers --- I suspect most of the metrics we want to expose can be determined without actually looking at the body, just the request headers & the _length_ of the streaming body.

> I don't know how Linkerd handles the buffer in case of Kafka traffic.

We'll have to answer this question as we build support for Kafka into the proxy :)